libndt.cpp

```diff
@@ -127,32 +128,28 @@ static std::string libndt_perror(Err err) noexcept {
   return rv;
 }
 
-#define EMIT_WARNING(statements)                         \
-  do {                                                   \
-    if (impl->settings.verbosity >= verbosity_warning) { \
-      std::stringstream ss;                              \
-      ss << statements;                                  \
-      on_warning(ss.str());                              \
-    }                                                    \
+// Generic macro for emitting logs. We lock the mutex when logging because
+// some log messages are emitted by background threads. Accessing the verbosity
+// is constant and verbosity does not change throughout the Client lifecycle,
+// hence it is a safe thing to do. Usually you probably don't want to log like
+// crazy, hence it's probably okay to use a mutex in this macro.
+#define EMIT_LOG_EX(client, level, statements)             \
+  do {                                                     \
+    if (client->get_verbosity() >= verbosity_##level) {    \
+      std::unique_lock<std::mutex> _{client->get_mutex()}; \
+      std::stringstream ss;                                \
+      ss << statements;                                    \
+      client->on_##level(ss.str());                        \
```

Comment:
So, after re-reading the code, it seems to me the `mutex` is justified by the desired to guarantee upstream that a specific `log_callback` cannot re-enter. However, this is probably the responsibility of the user, not mine.

Also, I am wondering whether I should make the inner verbosity level atomic. (In all fairness the accessor is constant and it is not possible to change the value during the test.)

Cc: @pboothe

Reply:
Something about introducing the possibility of mutex-based deadlock during LOG() calls makes me worry. It is likely fine, but I have gotten multi-threaded stuff wrong so many times that I mistrust it here.

Reply:
Thanks! I do not think that currently `LOG` calls `LOG`, but a derived class may be working differently and hence that could happen. And it would be very subtle. As a result, I have created #68 to remember to remove the mutex and #69 to make the settings explicitly `const`. Since above I said that "the accessor is constant" but the underlying settings object is not. Given that it is not supposed to change after being set, it's much better (especially from the point of view of helping the compiler to generate correct code) to store a `const Settings` internally.